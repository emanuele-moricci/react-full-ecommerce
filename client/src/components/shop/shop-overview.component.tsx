import { useSelector } from "react-redux";
import { selectCollections } from "src/redux/shop/shop.selectors";

import ShopItem from "src/components/shop/shop-item/shop-item.component";

import * as Styled from "./shop-overview.styles";

const ShopOverview = (): JSX.Element => {
  const collections = useSelector(selectCollections);

  return (
    <Styled.ShopOverview>
      {collections.map((collection) => (
        <ShopItem key={collection.id} {...collection} />
      ))}
    </Styled.ShopOverview>
  );
};

export default ShopOverview;
