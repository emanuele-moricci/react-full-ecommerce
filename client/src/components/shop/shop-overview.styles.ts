import styled from "styled-components";

export const ShopOverview = styled.div`
  display: flex;
  flex-direction: column;
`;
