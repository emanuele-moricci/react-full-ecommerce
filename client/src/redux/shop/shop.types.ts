export const ShopActionTypes = {
  FETCH_COLLECTIONS_START: "FETCH_COLLECTIONS_START",
  FETCH_COLLECTIONS_SUCCESS: "FETCH_COLLECTIONS_SUCCESS",
  FETCH_COLLECTIONS_FAILURE: "FETCH_COLLECTIONS_FAILURE",
};

export type Item = {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
};

export type Collection = {
  id: string;
  title: string;
  routeName: string;
  items: Item[];
};

export type CollectionList = {
  [key: string]: Collection;
};
