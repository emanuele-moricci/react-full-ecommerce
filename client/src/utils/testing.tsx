import { FC } from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";

export const fakePayloadAction = { type: "FAKE_PAYLOAD" };

export const getMockedStoreRenderedComponent = (state: any, Component: FC) => {
  const mockStore = configureStore();
  const store = mockStore(state);

  return renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Component />
      </BrowserRouter>
    </Provider>
  );
};
