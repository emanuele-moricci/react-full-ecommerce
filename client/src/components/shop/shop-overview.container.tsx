import { FC } from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectIsFetching } from "src/redux/shop/shop.selectors";

import ShopOverview from "./shop-overview.component";

import withSpinner from "src/components/HOC/with-spinner/with-spinner.component";

const mapStateToProps = createStructuredSelector({
  isLoading: selectIsFetching,
});

const ShopOverviewContainer = compose(
  connect(mapStateToProps),
  withSpinner
)(ShopOverview);

export default ShopOverviewContainer as FC;
