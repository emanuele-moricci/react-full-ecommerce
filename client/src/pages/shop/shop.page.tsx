import CollectionsOverviewContainer from "src/components/collection/collections-overview/collections-overview.container";

import * as Styled from "./shop.styles";

const Shop = (): JSX.Element => {
  return (
    <Styled.ShopPage>
      <CollectionsOverviewContainer />
    </Styled.ShopPage>
  );
};

export default Shop;
