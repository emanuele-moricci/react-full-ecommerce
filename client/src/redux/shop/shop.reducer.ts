import { AnyAction } from "redux";
import { ShopActionTypes } from "./shop.types";

const INITIAL_STATE = {
  collections: null,
  isFetching: false,
  errorMessage: null,
};

const shopReducer = (state = INITIAL_STATE, { type, payload }: AnyAction) => {
  switch (type) {
    case ShopActionTypes.FETCH_COLLECTIONS_START:
      return {
        ...state,
        isFetching: true,
      };
    case ShopActionTypes.FETCH_COLLECTIONS_SUCCESS:
      return {
        ...state,
        collections: payload,
        isFetching: false,
      };
    case ShopActionTypes.FETCH_COLLECTIONS_FAILURE:
      return {
        ...state,
        errorMessage: payload,
        isFetching: false,
      };
    default:
      return state;
  }
};

export default shopReducer;
