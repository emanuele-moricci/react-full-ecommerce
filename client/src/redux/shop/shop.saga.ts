import {
  collection,
  DocumentData,
  getDocs,
  QuerySnapshot,
} from "firebase/firestore";
import {
  convertCollectionsSnapshotToMap,
  firestore,
} from "src/db/firebase.utils";

import {
  all,
  call,
  takeLatest,
  put,
  CallEffect,
  PutEffect,
} from "redux-saga/effects";

import { CollectionList } from "./shop.types";
import { shopActions } from "./shop.slice";

/* ∨∨∨∨ START FUNCTIONS ∨∨∨∨ */

export function* shopSagas() {
  yield all([call(onFetchCollectionsStart)]);
}

export function* onFetchCollectionsStart() {
  yield takeLatest(
    shopActions.fetchCollectionsStart.type,
    fetchCollectionsAsync
  );
}

/* ^^^^ START FUNCTIONS ^^^^ */

export function* fetchCollectionsAsync(): Generator<
  Promise<QuerySnapshot<DocumentData>> | CallEffect<CollectionList> | PutEffect
> {
  try {
    const collectionRef = collection(firestore, "collections");
    const snapshot = (yield getDocs(
      collectionRef
    )) as QuerySnapshot<DocumentData>;

    const collectionsMap = (yield call(
      convertCollectionsSnapshotToMap,
      snapshot
    )) as CollectionList;

    yield put(shopActions.fetchCollectionsSuccess(collectionsMap));
  } catch (error) {
    const msg = error instanceof Error ? error.message : "";

    yield put(shopActions.fetchCollectionsFailure(msg));
  }
}
