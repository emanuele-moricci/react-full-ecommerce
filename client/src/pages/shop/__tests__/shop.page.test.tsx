import { getMockedStoreRenderedComponent } from "src/utils/testing";

import { initialState } from "../state/shop.slice";

import Shop from "src/pages/shop/shop.page";

describe("<Shop />", () => {
  it("Should match the snapshot", () => {
    const component = getMockedStoreRenderedComponent(
      { shop: initialState },
      Shop
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
