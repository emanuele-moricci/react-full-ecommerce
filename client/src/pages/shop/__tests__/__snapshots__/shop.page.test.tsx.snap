// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shop /> Should match the snapshot 1`] = `
<div
  className="sc-iqseJM"
>
  <div
    className="sc-pVTFL dMzYpF"
  />
</div>
`;
